### PR TITLE
Support colors and symbols for select one from map

### LINKED ISSUE #5278: Support colors for select from map markers
- [x] Given that I have a form definition with a select from map
  And that a select option has a value set for a property with name `marker-color` 
  And that the value is valid based on the criteria below
  When I open the select from map question
  Then the marker for that option should be the specified color

- [x] Given that I have a form definition with a select from map
  And that a select option has no value set for a property with name `marker-color` 
  When I open the select from map question
  Then the marker for that option should be the default color

- [x] Given that I have a form definition with a select from map
  And that a select option has a value set for a property with name `marker-color` 
  And that the value is NOT valid based on the criteria below
  When I open the select from map question
  Then the marker for that option should be the default color

Color criteria from [simple style spec](https://github.com/mapbox/simplestyle-spec/tree/master/1.1.0):
```
// Colors can be in short form:
//   "#ace"
// or long form
//   "#aaccee"
// with or without the # prefix.
// Colors are interpreted the same as in CSS,
// in #RRGGBB and #RGB order.
```

Select options may come from a geojson file, a csv file, or be internal to the form.

### LINKED ISSUE #5279: Support alphanumeric markers for select one from map
- [x] Given that I have a form definition with a select from map
And that a select option has a value set for a property with name `marker-symbol`
And that the value is a single character
When I open the select from map question
Then the marker for that option should be the place marker icon with the character centered on its pin

(Roughly Screen Shot 2022-09-07 at 1 22 24 PM)

- [x] Given that I have a form definition with a select from map
And that a select option has a value set for a property with name `marker-symbol`
And that the value is NOT a single character
When I open the select from map question
Then the marker for that option should be the standard place marker icon

- [x] Given that I have a form definition with a select from map
And that a select option has a value set for a property with name `marker-symbol`
And that it has no value
When I open the select from map question
Then the marker for that option should be the standard place marker icon

See [simple style spec](https://github.com/mapbox/simplestyle-spec/tree/master/1.1.0).

Select options may come from a geojson file, a csv file, or be internal to the form.

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #5278
Closes #5279

#### What has been done to verify that this works as intended?
I've tested the change manually.

#### Why is this the best possible solution? Were any other approaches considered?
I don't think there were any other (better approach). I reused the implementation (abstractions) we already had. The only think of note is the refactoring. Previously we passed marker details in many places not in one object but as a set of params. I changed it to make it easier and clearer and it's visible in commits like:
- https://github.com/getodk/collect/pull/5280/commits/93e90ebde760f111c756641107bd0dc6cf442b00
- https://github.com/getodk/collect/pull/5280/commits/d765e2ee2f1dedf5037511df65798578cb86e0da
- https://github.com/getodk/collect/pull/5280/commits/1152adea53428c8a73bfa918b6899f701d042f58

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Please test adding markers to maps (all providers). Not only select one from map should be tested but also geo widgets.

#### Do we need any specific form for testing your changes? If so, please attach one.
I used:  [form.zip](https://github.com/getodk/collect/files/9560979/form.zip)

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
Yes https://github.com/getodk/docs/issues/1488

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by grzesiek2010** [comment]
I've decided to create one pull request for the two issues since they are related and it will save us some time not only during implementing the feature but also during testing.

There are some things to discus:
1. So far we have used marker icons with a transparent hole it's something that doesn't look well of course with symbols so I had to add a new marker (without that hole) for markers that we use in select one from map. Please take a look at it and let me know if there is something you do not like about them. They seem to be a bit bigger but I can change it if it's a problem (we might need to first check how it looks like on different screens)
2. As I said above the new marker without a hole is used only for select one from map no matter if there is a symbol specified or not. There are a couple of options here:
     - We can leave it as is (use the new icon for select one form map and the old ones with holes in the map of forms)
     - We can use the new icon without a hole only if there is a symbol specified and in all the other cases use the old ones.
     - We can use the new icon everywhere 
I'm not sure, I think that it looks a bit weird to have that icon without a hole even if there is no symbol so maybe the second option would be the one I like more...
3. As I understand there is no need to display that new properties (marker-color, marker-symbol) in the bottom sheet? They should be only used to customize markers right?  

| Current version | New markers no color/symbol | New markers with color/symbol |
| ------------- | ------------- | ------------- |
| Screenshot_1662719740 |  Screenshot_1662719592 | Screenshot_1662719611 |

@seadowg @lognaturel please let me know what you think.

**Comment by seadowg** [comment]
> So far we have used marker icons with a transparent hole it's something that doesn't look well of course with symbols so I had to add a new marker (without that hole) for markers that we use in select one from map.

Looks good!

> We can use the new icon everywhere

It feels like it would be good to keep the marker shape as consistent as possible across our maps. I'd vote for this option I think.

> As I understand there is no need to display that new properties (marker-color, marker-symbol) in the bottom sheet? They should be only used to customize markers right?

Unless @lognaturel has a good reason, I don't think we should do the extra work to filter out properties.

**Comment by grzesiek2010** [comment]
> Unless @lognaturel has a good reason, I don't think we should do the extra work to filter out properties.

I thought it was the expected behavior so I've already done that extra work. If you agree we can keep it.

**Comment by grzesiek2010** [comment]
This is how it looks now:

| Emoji / brown color  | W symbol / red color | No symbol / no color | No symbol / green color | P symbol / white color |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| Screenshot_1663101513 | Screenshot_1663101517 | Screenshot_1663101520  | Screenshot_1663101524 | Screenshot_1663101528 |

File: geo/src/main/java/org/odk/collect/geo/selection/SelectionMapFragment.kt [hunk 1]
```diff
@@ -272,7 +278,7 @@ class SelectionMapFragment(
                 } else {
                     map.zoomToPoint(MapPoint(item.latitude, item.longitude), true)
                 }
-                map.setMarkerIcon(featureId, item.largeIcon)
+                map.setMarkerIcon(featureId, MarkerIconDescription(item.largeIcon, item.color, item.symbol))
```
Comment (seadowg) [inline geo/src/main/java/org/odk/collect/geo/selection/SelectionMapFragment.kt]: As far as I can see, there aren't any tests to check that the map markers are set up with the correct color and symbol (or that they are retained when switching icon).

File: collect_app/src/main/java/org/odk/collect/android/widgets/items/SelectOneFromMapDialogFragment.kt [hunk 2]
```diff
@@ -192,19 +192,24 @@ internal class SelectChoicesMapData(
...
+                val markerColor = selectChoice.additionalChildren.firstOrNull { it.first == "marker-color" }?.second
+                val markerSymbol = selectChoice.additionalChildren.firstOrNull { it.first == "marker-symbol" }?.second
+
                 list + MappableSelectItem.WithAction(
                     index.toLong(),
                     latitude,
                     longitude,
-                    R.drawable.ic_map_marker_24dp,
-                    R.drawable.ic_map_marker_48dp,
+                    if (markerSymbol == null) R.drawable.ic_map_marker_with_hole_small else R.drawable.ic_map_marker_small,
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/widgets/items/SelectOneFromMapDialogFragment.kt:202]: There's no test that checks this change (choosing the icon based on the marker symbol)

File: collect_app/src/main/java/org/odk/collect/android/widgets/items/SelectOneFromMapDialogFragment.kt [hunk 3]
```diff
@@ -192,19 +192,24 @@ internal class SelectChoicesMapData(
...
                     prompt.getSelectChoiceText(selectChoice),
                     properties,
                     MappableSelectItem.IconifiedText(
                         R.drawable.ic_save,
                         resources.getString(R.string.select_item)
                     ),
-                    selectChoice.index == selectedIndex
+                    selectChoice.index == selectedIndex,
+                    markerColor,
+                    markerSymbol
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/widgets/items/SelectOneFromMapDialogFragment.kt:212]: There's no test to check whether the `markerSymol` and `markerColor` are extracted from the choice's additional children and passed in correctly here. We should also have tests for cases where "marker-color" or "marker-symbol" don't exist.

File: maps/src/main/java/org/odk/collect/maps/MarkerIconDescription.kt [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+    fun getSymbol(): String? = symbol?.let {
+        if (it.isBlank()) {
+            null
+        } else {
+            StringUtils.firstCharacterOrEmoji(it).uppercase(Locale.US)
+        }
+    }
+
+    override fun hashCode(): Int {
```
Comment (seadowg) [inline maps/src/main/java/org/odk/collect/maps/MarkerIconDescription.kt]: Not seeing why `hashCode` is overridden here? I get that there would need to be a custom implementation to maintain the behaviour we lose from moving away from `data class`, but I can't find anything that requires that behaviour. Additionally, even those this doesn't breach the `hashCode`/`equals` relationship (objects that are equal would have the same `hashCode` here), it does feel weird to have a smart `hashCode` implementation but leave `equals` using basic object equality as it massively increases the cases where two objects can have the same `hashCode` but be "unequal".

**Review by seadowg — CHANGES_REQUESTED** [review]
I'll make changes based on my comments as @grzesiek2010 is away, but wanted to make sure it was clear what changes I was requesting to explain my follow ups.

**Comment by seadowg** [comment]
@getodk/testers this is ready for manual testing as all the remaining requested changes are just adding automated tests.

File: maps/src/main/java/org/odk/collect/maps/MarkerIconDescription.kt (hunk 4 above)
Comment (seadowg) [reply maps/src/main/java/org/odk/collect/maps/MarkerIconDescription.kt]: I've removed this and all tests are still passing.

**Comment by srujner** [comment]
I noticed one strange behavior on every device:

https://user-images.githubusercontent.com/85225519/192499675-e1ebc40f-3086-4dcc-a8ce-a47b465bcdda.mp4

The "Berlin" point doesn't have "marker-color:" information in info, maybe that's the reason. 
I was able to reproduce it in geojson form to.

**Comment by srujner** [comment]
@seadowg @lognaturel One more thing to discuss:
The "P" point is barely visible on white background. Maybe adding some black outline to points wouldn't be a bad idea? 
So that they do not blend in with the background.

Screenshot_20220927-110314(1)

**Comment by seadowg** [comment]
> I noticed one strange behavior on every device:

This definitely looks like a bug. I'll investigate.

> The "P" point is barely visible on white background. Maybe adding some black outline to points wouldn't be a bad idea?
So that they do not blend in with the background.

Nice find. Unless there's something I'm missing in the spec, I think this is something we should consider "user error" - it's probably a bad idea to use white markers. Does that sound fair @lognaturel? If not, I think we should write up a separate issue for the extra hand holding of special cases for hard to see markers.

**Comment by seadowg** [comment]
@srujner could you confirm you're seeing the problem with markers without `marker-color` using the wrong color with the latest version of the PR please?

**Comment by srujner** [comment]
@seadowg I can confirm to reproduce the issue with the latest version. Sometimes you need to select the markers first or change map source and then click multiple times on markers to reproduce it.

**Comment by seadowg** [comment]
@srujner could you send me the form you're using to reproduce?

**Comment by lognaturel** [comment]
> I think this is something we should consider "user error" 

Yes, I agree. We could eventually consider an outline for all markers -- this issue is not specific to the new features.

**Comment by srujner** [comment]
@seadowg It's the one attached by Grzegorz
[form(1).zip](https://github.com/getodk/collect/files/9662743/form.1.zip)

**Comment by seadowg** [comment]
@srujner ok that should be fixed now.

@grzesiek2010 this was a weird one! It turns out that we need to be calling [`mutate`](https://developer.android.com/reference/android/graphics/drawable/Drawable.html#mutate()) on `Drawable` objects we want to modify, or we end up with some state being shared between resources with the same ID (the docs explain this better than I would). This actually might explain some issues we've seen with day/night theme switching - although that will be most likely due to a lack of `mutate` calls in framework code.

**Comment by srujner** [comment]
**Tested with Success!**
Verified on Android: 10, 12 and 13 

Verified cases:
    • Regression checks for GeoPoint, GeoTrace, GeoShape widgets; 
    • Select Map forms;
    • Adding markers to maps;
    • Offline Layers; 
    • Google Maps, Mapbox, OSM and all map styles; 
    • Collecting Location in background; 
    • View Submissions on map; 
    • Screen rotation, minimize app, using backward button,

**Review by seadowg — APPROVED** [review]

**Pull request merged**